### PR TITLE
Refactor system contracts

### PR DESCRIPTION
**Pull request by pattyshack**
1. Renamed ContractFunctionInvoker to SystemContracts
2. Converted functions in contractFunctionInvocations.go into SystemContracts
   methods
3. Add AddressFromChain to ContractionFunctionSpec
4. Add SystemContracts to environment object

Note: I'll rename and concat the files when I move them into the environment
package.

### CODE REVIEW

File: fvm/contractFunctionInvocations.go [hunk 1]
```diff
@@ -21,15 +31,13 @@ var deductTransactionFeeSpec = ContractFunctionSpec{
 
 // InvokeDeductTransactionFeesContract executes the fee deduction contract on
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:32]:
```suggestion
// DeductTransactionFees executes the fee deduction contract on
```

File: fvm/contractFunctionInvocations.go [hunk 2]
```diff
@@ -50,14 +59,12 @@ var setupNewAccountSpec = ContractFunctionSpec{
 
 // InvokeSetupNewAccountContract executes the new account setup contract on
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:60]:
```suggestion
// SetupNewAccount executes the new account setup contract on
```

File: fvm/contractFunctionInvocations.go [hunk 3]
```diff
@@ -66,68 +73,65 @@ func InvokeSetupNewAccountContract(
...
-	FunctionName: systemcontracts.ContractStorageFeesFunction_defaultTokenAvailableBalance,
+	AddressFromChain: ServiceAddress,
+	LocationName:     systemcontracts.ContractStorageFees,
+	FunctionName:     systemcontracts.ContractStorageFeesFunction_defaultTokenAvailableBalance,
 	ArgumentTypes: []sema.Type{
 		&sema.AddressType{},
 	},
 }
 
 // InvokeAccountAvailableBalanceContract executes the get available balance
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:84]:
```suggestion
// AccountAvailableBalance executes the get available balance
```

File: fvm/contractFunctionInvocations.go [hunk 4]
```diff
@@ -66,68 +73,65 @@ func InvokeSetupNewAccountContract(
...
-	FunctionName: systemcontracts.ContractServiceAccountFunction_defaultTokenBalance,
+	AddressFromChain: ServiceAddress,
+	LocationName:     systemcontracts.ContractServiceAccount,
+	FunctionName:     systemcontracts.ContractServiceAccountFunction_defaultTokenBalance,
 	ArgumentTypes: []sema.Type{
 		sema.PublicAccountType,
 	},
 }
 
 // InvokeAccountBalanceContract executes the get available balance contract
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:106]:
```suggestion
// AccountBalance executes the get available balance contract
```

File: fvm/contractFunctionInvocations.go [hunk 5]
```diff
@@ -66,68 +73,65 @@ func InvokeSetupNewAccountContract(
...
-	FunctionName: systemcontracts.ContractStorageFeesFunction_calculateAccountCapacity,
+	AddressFromChain: ServiceAddress,
+	LocationName:     systemcontracts.ContractStorageFees,
+	FunctionName:     systemcontracts.ContractStorageFeesFunction_calculateAccountCapacity,
 	ArgumentTypes: []sema.Type{
 		&sema.AddressType{},
 	},
 }
 
 // InvokeAccountStorageCapacityContract executes the get storage capacity
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:128]:
```suggestion
// AccountStorageCapacity executes the get storage capacity
```

File: fvm/contractFunctionInvocations.go [hunk 6]
```diff
@@ -136,19 +140,19 @@ func InvokeAccountStorageCapacityContract(
 
 // InvokeAccountsStorageCapacity prepares a function that calls get storage capacity on the storage fees contract
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:141]:
```suggestion
// AccountsStorageCapacity prepares a function that calls get storage capacity on the storage fees contract
```

File: fvm/contractFunctionInvocations.go [hunk 7]
```diff
@@ -175,14 +179,12 @@ var useContractAuditVoucherSpec = ContractFunctionSpec{
 
 // InvokeUseContractAuditVoucherContract executes the use a contract
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvocations.go:180]:
```suggestion
// UseContractAuditVoucher executes the use a contract
```

File: fvm/contractFunctionInvoker.go [hunk 8]
```diff
@@ -15,61 +15,61 @@ import (
...
-	FunctionName  string
-	ArgumentTypes []sema.Type
+	AddressFromChain func(flow.Chain) flow.Address
+	LocationName     string
+	FunctionName     string
+	ArgumentTypes    []sema.Type
 }
 
-type ContractFunctionInvoker struct {
+type SystemContracts struct {
```
Comment (zhangchiqing) [inline fvm/contractFunctionInvoker.go:24]:
```suggestion
// SystemContracts provides methods for invoking system contract functions as service account.
type SystemContracts struct {
```

**Review by zhangchiqing — APPROVED** [review]

**Review by SaveTheRbtz — APPROVED** [review]

**Comment by pattyshack** [comment]
bors merge

**Pull request merged**